examples/qabot/train.py

```diff
@@ -0,0 +1,162 @@
+            q_vec = words_text_to_fixed_seqlen_vec(wv,q,q_seq_length)
+            tq.copy_from_numpy(q_vec)
+
+            a_pos_vec = words_text_to_fixed_seqlen_vec(wv,a_pos,a_seq_length)
+            # prepare for <q, a+, a+> input
+            ta.copy_from_numpy(np.array([a_pos_vec]*2))
+            true_score, l = m(tq,ta)
+
+            a_neg_vecs = [words_text_to_fixed_seqlen_vec(wv,a_neg,a_seq_length) for a_neg in a_negs]
+
+            # prepare for triplets <q, a-, a-> input
+            while len(a_neg_vecs) > 1:
+                a_vec=[]
+                a_vec.append(a_neg_vecs.pop(0))
+                a_vec.append(a_neg_vecs.pop(0))
+                ta.copy_from_numpy(np.array(a_vec))
+                score, l = m(tq,ta)
+                scores.extend(score)
+
+            max_neg = np.max(np.array([tensor.to_numpy(s) for s in scores]).flatten())
+            if max_neg < tensor.to_numpy(true_score[0])[0]:
+                top1hit+=1
+
+        print("training top 1 hit accuracy: ", top1hit/trials)
+
+
+def test(m, tq, ta, id_to_word, label_to_ans_text, wv):
+    print("testing")
+    test_data = parse_test_file('./V2/InsuranceQA.question.anslabel.token.100.pool.solr.test.encoded', id_to_word, label_to_ans_text)
+    test_data = test_data[:10]  # run on n samples
+
+    m.eval()
+    top1hit=0
+    trials = len(test_data)
+    for (q, labels, cands) in test_data:
+
+        q_vec = words_text_to_fixed_seqlen_vec(wv, q, q_seq_length)
+        tq.copy_from_numpy(np.array(q_vec))
+
+        cands_vec = [words_text_to_fixed_seqlen_vec(wv, label_to_ans_text[candidate_label], a_seq_length) for candidate_label in cands]
+
+        scores = []
+        # inference all candidates
+        # import pdb; pdb.set_trace()
+        while len(cands_vec) > 1:
+            a_vec=[]
+            a_vec.append(cands_vec.pop(0))
+            a_vec.append(cands_vec.pop(0))
+            ta.copy_from_numpy(np.array(a_vec))
+            score = m(tq,ta) # inference mode only return forward result
+            scores.extend(score)
+
+        # check correct from predict
+        true_idxs = [cands.index(l) for l in labels if l in cands]
+        pred_idx = np.argmax(np.array([tensor.to_numpy(s) for s in scores]).flatten())
+        if pred_idx in true_idxs:
+            top1hit += 1
+
+    print("testing top 1 hit accuracy: ", top1hit/trials)
+
+if __name__ == "__main__":
+    dev = device.create_cuda_gpu(set_default=False)
+
+    q_seq_length = 10
+    a_seq_length = 100
+    embed_size = 300
+    batch_size = 128
+    max_epoch = 30
+    hidden_size = 100
+
+    # build model
+    m = QAModel(hidden_size)
+    print("created qa model")
+    # m = MLP()
+    m.optimizer = opt.SGD()
+
+    tq = tensor.Tensor((1, q_seq_length, embed_size), dev, tensor.float32)
+    ta = tensor.Tensor((2, a_seq_length, embed_size), dev, tensor.float32)
+
+    tq.set_value(0.0)
+    ta.set_value(0.0)
+
+    m.compile([tq, ta], is_train=True, use_graph=False, sequential=False)
```

Comment:
If the graph is enabled, can we train the model correctly?

Reply:
setting use_graph to True gives:
```
Traceback (most recent call last):
  File "/root/singa/examples/qabot/train.py", line 184, in <module>
    training(m, train_data, max_epoch)
  File "/root/singa/examples/qabot/train.py", line 126, in training
    score, l = m(tq, ta)
  File "build/python/singa/model.py", line 210, in __call__
    return self.train_one_batch(*input, **kwargs)
  File "build/python/singa/model.py", line 63, in wrapper
    self._results = func(self, *args, **kwargs)
  File "/root/singa/examples/qabot/model.py", line 65, in train_one_batch
    self.optimizer.backward_and_update(loss)
  File "/usr/local/lib/python3.6/dist-packages/deprecated/classic.py", line 281, in wrapper_function
    return wrapped_(*args_, **kwargs_)
  File "build/python/singa/opt.py", line 320, in backward_and_update
    super(SGD, self).__call__(loss)
  File "build/python/singa/opt.py", line 103, in __call__
    self.call(loss)
  File "build/python/singa/opt.py", line 110, in call
    self.apply(p.name, p, g)
  File "build/python/singa/opt.py", line 270, in apply
    singa.Axpy(minus_lr.data, param_grad.data, param_value.data)
TypeError: in method 'Axpy', argument 1 of type 'float'
```